File: relay/hub/signal.js
// vim: ts=4:sw=4:expandtab

'use strict';

const ProvisioningCipher = require('../provisioning_cipher');
const errors = require('../errors');
const fetch = require('./fetch');
const libsignal = require('libsignal');
const protobufs = require('../protobufs');
const storage = require('../storage');

const SIGNAL_URL_CALLS = {
    accounts: "/v1/accounts",
    devices: "/v1/devices",
    keys: "/v2/keys",
    messages: "/v1/messages",
    attachment: "/v1/attachments"
};

const SIGNAL_HTTP_MESSAGES = {
    401: "Invalid authentication or invalidated registration",
    403: "Invalid code",
    404: "Address is not registered",
    413: "Server rate limit exceeded",
    417: "Address already registered"
};

const lastResortKeyId = 0xdeadbeef & ((2 ** 31) - 1); // Must fit inside signed 32bit int.


class SignalClient {

    constructor(username, password, url) {
        this.url = url;
        this.username = username;
        this.password = password;
        this.attachment_id_regex = RegExp("^https://.*/(\\d+)?");
    }

    static async factory() {
        const url = await storage.getState('serverUrl');
        const username = await storage.getState('username');
        const password = await storage.getState('password');
        return new this(username, password, url);
    }

    async linkDevice(uuid, pubKey, options) {
        options = options || {};
        const provisionResp = await this.request({
            call: 'devices',
            urlParameters: '/provisioning/code'
        });
        const ourIdent = await storage.getOurIdentity();
        const pMessage = new protobufs.ProvisionMessage();
        pMessage.identityKeyPrivate = ourIdent.privKey;
        pMessage.addr = await storage.getState('addr');
        pMessage.userAgent = options.userAgent || 'librelay-web';
        pMessage.provisioningCode = provisionResp.verificationCode;
        const provisioningCipher = new ProvisioningCipher();
        const pEnvelope = await provisioningCipher.encrypt(pubKey, pMessage);
        const resp = await this.fetch('/v1/provisioning/' + uuid, {
            method: 'PUT',
            json: {
                body: pEnvelope.toString('base64') // XXX probably have to finish()/encode() this thing.
            }
        });
        if (!resp.ok) {
            // 404 means someone else handled it already.
            if (resp.status !== 404) {
                throw new Error(await resp.text());
            }
        }
    }

    async refreshPreKeys(minLevel=10, fill=100) {
        const preKeyCount = await this.getMyKeys();
        const lastResortKey = await storage.loadPreKey(lastResortKeyId);
        if (preKeyCount <= minLevel || !lastResortKey) {
            // The server replaces existing keys so just go to the hilt.
            console.info("Refreshing pre-keys...");
            await this.registerKeys(await this.generateKeys(fill));
        }
    }

    async generateKeys(count=100, progressCallback) {
        if (typeof progressCallback !== 'function') {
            progressCallback = undefined;
        }
        const startId = await storage.getState('maxPreKeyId') || 1;
        const signedKeyId = await storage.getState('signedKeyId') || 1;

        if (typeof startId != 'number') {
            throw new Error('Invalid maxPreKeyId');
        }
        if (typeof signedKeyId != 'number') {
            throw new Error('Invalid signedKeyId');
        }

        let lastResortKey = await storage.loadPreKey(lastResortKeyId);
        if (!lastResortKey) {
            // Last resort key only used if our prekey pool is drained faster than
            // we refresh it.  This prevents message dropping at the expense of
            // forward secrecy impairment.
            const pk = await libsignal.KeyHelper.generatePreKey(lastResortKeyId);
            await storage.storePreKey(lastResortKeyId, pk.keyPair);
            lastResortKey = pk.keyPair;
        }
        const ourIdent = await storage.getOurIdentity();
        const result = {
            preKeys: [],
            identityKey: ourIdent.pubKey,
            lastResortKey: {
                keyId: lastResortKeyId,
                publicKey: lastResortKey.pubKey
            }
        };
        for (let keyId = startId; keyId < startId + count; ++keyId) {
            const preKey = await libsignal.KeyHelper.generatePreKey(keyId);
            await storage.storePreKey(preKey.keyId, preKey.keyPair);
            result.preKeys.push({
                keyId: preKey.keyId,
                publicKey: preKey.keyPair.pubKey
            });
            if (progressCallback) {
                progressCallback(keyId - startId);
            }
        }
        const sprekey = await libsignal.KeyHelper.generateSignedPreKey(ourIdent, signedKeyId);
        await storage.storeSignedPreKey(sprekey.keyId, sprekey.keyPair);
        result.signedPreKey = {
            keyId: sprekey.keyId,
            publicKey: sprekey.keyPair.pubKey,
            signature: sprekey.signature
        };
        await storage.removeSignedPreKey(signedKeyId - 2);
        await storage.putState('maxPreKeyId', startId + count);
        await storage.putState('signedKeyId', signedKeyId + 1);
        return result;
    }

    /* BREAK api signal code.... */

    authHeader(username, password) {
        const token = Buffer.from(username + ':' + password).toString('base64');
        return 'Basic ' + token;
    }

    validateResponse(response, schema) {
        try {
            for (var i in schema) {
                switch (schema[i]) {
                    case 'object':
                    case 'string':
                    case 'number':
                        if (typeof response[i] !== schema[i]) {
                            return false;
                        }
                        break;
                }
            }
        } catch(ex) {
            return false;
        }
        return true;
    }

    async request(param) {
        if (!param.urlParameters) {
            param.urlParameters = '';
        }
        const path = SIGNAL_URL_CALLS[param.call] + param.urlParameters;
        const headers = new fetch.Headers();
        if (param.username && param.password) {
            headers.set('Authorization', this.authHeader(param.username, param.password));
        }
        let resp;
        try {
            resp = await this.fetch(path, {
                method: param.httpType || 'GET',
                json: param.jsonData,
                headers
            });
        } catch(e) {
            /* Fetch throws a very boring TypeError, throw something better.. */
            throw new errors.NetworkError(`${e.message}: ${param.call}`);
        }
        let resp_content;
        if ((resp.headers.get('content-type') || '').startsWith('application/json')) {
            resp_content = await resp.json();
        } else {
            resp_content = await resp.text();
        }
        if (!resp.ok) {
            const e = new errors.ProtocolError(resp.status, resp_content);
            if (SIGNAL_HTTP_MESSAGES.hasOwnProperty(e.code)) {
                e.message = SIGNAL_HTTP_MESSAGES[e.code];
            } else {
                e.message = `Status code: ${e.code}`;
            }
            throw e;
        }
        if (resp.status !== 204) {
            if (param.validateResponse &&
                !this.validateResponse(resp_content, param.validateResponse)) {
                throw new errors.ProtocolError(resp.status, resp_content);
            }
            return resp_content;
        }
    }

    async fetch(urn, options) {
        /* Thin wrapper to augment json and auth support. */
        options = options || {};
        options.headers = options.headers || new fetch.Headers();
        if (!options.headers.has('Authorization')) {
            if (this.username && this.password) {
                options.headers.set('Authorization', this.authHeader(this.username, this.password));
            }
        }
        return await fetch(`${this.url}/${urn.replace(/^\//, '')}`, options);
    }

    async getDevices() {
        const data = await this.request({call: 'devices'});
        return data && data.devices;
    }

    async registerKeys(genKeys) {
        var jsonData = {};
        jsonData.identityKey = genKeys.identityKey.toString('base64');
        jsonData.signedPreKey = {
            keyId: genKeys.signedPreKey.keyId,
            publicKey: genKeys.signedPreKey.publicKey.toString('base64'),
            signature: genKeys.signedPreKey.signature.toString('base64')
        };
        jsonData.preKeys = [];
        var j = 0;
        for (var i in genKeys.preKeys) {
            jsonData.preKeys[j++] = {
                keyId: genKeys.preKeys[i].keyId,
                publicKey: genKeys.preKeys[i].publicKey.toString('base64')
            };
        }
        // Newer generation servers don't expect this BTW.
        jsonData.lastResortKey = {
            keyId: genKeys.lastResortKey.keyId,
            publicKey: genKeys.lastResortKey.publicKey.toString('base64')
        };
        return await this.request({
            call: 'keys',
            httpType: 'PUT',
            jsonData
        });
    }

    async getMyKeys() {
        const res = await this.request({
            call: 'keys',
            validateResponse: {count: 'number'}
        });
        return res.count;
    }

    async getKeysForAddr(addr, deviceId) {
        if (deviceId === undefined) {
            deviceId = "*";
        }
        const res = await this.request({
            call: 'keys',
            urlParameters: "/" + addr + "/" + deviceId,
            validateResponse: {identityKey: 'string', devices: 'object'}
        });
        if (res.devices.constructor !== Array) {
            throw new TypeError("Invalid response");
        }
        res.identityKey = Buffer.from(res.identityKey, 'base64');
        for (const device of res.devices) {
            if (!this.validateResponse(device, {signedPreKey: 'object'}) ||
                !this.validateResponse(device.signedPreKey, {publicKey: 'string', signature: 'string'})) {
                throw new TypeError("Invalid signedPreKey");
            }
            if (device.preKey) {
                if (!this.validateResponse(device, {preKey: 'object'}) ||
                    !this.validateResponse(device.preKey, {publicKey: 'string'})) {
                    throw new TypeError("Invalid preKey");
                }
                device.preKey.publicKey = Buffer.from(device.preKey.publicKey, 'base64');
            }
            device.signedPreKey.publicKey = Buffer.from(device.signedPreKey.publicKey, 'base64');
            device.signedPreKey.signature = Buffer.from(device.signedPreKey.signature, 'base64');
        }
        return res;
    }

    async sendMessages(destination, messages, timestamp) {
        return await this.request({
            call: 'messages',
            httpType: 'PUT',
            urlParameters: '/' + destination,
            jsonData: {messages, timestamp}
        });
    }

    async getAttachment(id) {
        // XXX Build in retry handling...
        const response = await this.request({
            call: 'attachment',
            urlParameters: '/' + id,
            validateResponse: {location: 'string'}
        });
        const headers = new fetch.Headers({
            'Content-Type': 'application/octet-stream',
        });
        const attachment = await fetch(response.location, {headers});
        if (!attachment.ok) {
            const msg = await attachment.text();
            console.error("Download attachement error:", msg);
            throw new Error('Download Attachment Error: ' + msg);
        }
        return await attachment.buffer();
    }

    async putAttachment(body) {
        // XXX Build in retry handling...
        const ptrResp = await this.request({call: 'attachment'});
        // Extract the id as a string from the location url
        // (workaround for ids too large for Javascript numbers)
        const match = ptrResp.location.match(this.attachment_id_regex);
        if (!match) {
            console.error('Invalid attachment url for outgoing message',
                          ptrResp.location);
            throw new TypeError('Received invalid attachment url');
        }
        const headers = new fetch.Headers({
            'Content-Type': 'application/octet-stream',
            'Content-Length': body.byteLength  // See: https://github.com/bitinn/node-fetch/issues/47
        });
        const dataResp = await fetch(ptrResp.location, {
            method: "PUT",
            headers,
            body
        });
        if (!dataResp.ok) {
            const msg = await dataResp.text();
            console.error("Upload attachement error:", msg);
            throw new Error('Upload Attachment Error: ' + msg);
        }
        return match[1];
    }

    getMessageWebSocketURL() {
        return [
            this.url.replace('https://', 'wss://').replace('http://', 'ws://'),
            '/v1/websocket/?login=', encodeURIComponent(this.username),
            '&password=', encodeURIComponent(this.password)].join('');
    }

    getProvisioningWebSocketURL () {
        return this.url.replace('https://', 'wss://').replace('http://', 'ws://') +
                                '/v1/websocket/provisioning/';
    }

    /* The GCM reg ID configures the data needed for the PushServer to wake us up
     * using google cloud messaging's Push Server (an exercise for the user) */
    async updateGcmRegistrationId(gcm_reg_id) {
        return await this.request({
            call: 'accounts',
            httpType: 'PUT',
            urlParameters: '/gcm',
            jsonData: {
                gcmRegistrationId: gcm_reg_id
            }
        });
    }
}

module.exports = SignalClient;

File: relay/hub/signal.py
import aiohttp
import asyncio
import logging
import re
from .. import protobufs
from .. import storage
from ..provisioning_cipher import ProvisioningCipher
from axolotl.util.keyhelper import KeyHelper
from base64 import b64decode, b64encode

store = storage.getStore()
logger = logging.getLogger(__name__)

SIGNAL_URL_CALLS = {
    "accounts": "/v1/accounts",
    "devices": "/v1/devices",
    "keys": "/v2/keys",
    "messages": "/v1/messages",
    "attachment": "/v1/attachments"
}

SIGNAL_HTTP_MESSAGES = {
    401: "Invalid authentication or invalidated registration",
    403: "Invalid code",
    404: "Address is not registered",
    413: "Server rate limit exceeded",
    417: "Address already registered"
}


class SignalClient(object):

    attachment_id_regex = re.compile("^https://.*/(\\d+)?")

    def __init__(self, username, password, url):
        self.url = url
        self.username = username
        self.password = password
        self._httpClient = aiohttp.ClientSession(read_timeout=30,
                                                 raise_for_status=True)

    def __del__(self):
        asyncio.get_event_loop().create_task(self._httpClient.close())
        self._httpClient = None

    @classmethod
    def factory(cls):
        url = store.getState('serverUrl')
        username = store.getState('username')
        password = store.getState('password')
        return cls(username, password, url)

    async def linkDevice(self, uuid, pubkey, userAgent='librelay-python'):
        provision_resp = await self.request(call='devices',
                                            urn='/provisioning/code')
        our_ident = store.getOurIdentity()
        pMessage = protobufs.ProvisionMessage()
        pMessage.identityKeyPrivate = our_ident.getPrivateKey().serialize()
        pMessage.addr = store.getState('addr')
        pMessage.userAgent = userAgent
        pMessage.provisioningCode = provision_resp.verificationCode
        provisioningCipher = ProvisioningCipher()
        pEnvelope = await provisioningCipher.encrypt(pubkey, pMessage)
        try:
            async with await self.fetch('/v1/provisioning/' + uuid, method='PUT',
                                        json={"body": b64encode(pEnvelope)}):
                pass
        except aiohttp.ClientResponseError as e:
            # 404 is okay, just means someone else handled it already.
            if e.status != 404:
                raise e

    async def refreshPreKeys(self, minLevel=10, fill=100):
        preKeyCount = await self.getMyKeys()
        if preKeyCount <= minLevel:
            # The server replaces existing keys so just go to the hilt.
            logger.info("Refreshing pre-keys...")
            await self.register_keys(await self.generate_keys(fill))

    async def generateKeys(self, count=100):
        startId = store.getState('maxPreKeyId') or 1
        signedKeyId = store.getState('signedKeyId') or 1
        ourIdent = store.getOurIdentity()
        result = {
            "identityKey": ourIdent.getPublicKey().serialize(),
            "preKeys": []
        }
        preKeys = KeyHelper.generatePreKeys(startId, count)
        for pk in preKeys:
            store.storePreKey(pk.getId(), pk)
            result['preKeys'].append({
                "keyId": pk.getId(),
                "publicKey": pk.getKeyPair().getPublicKey().serialize()
            })
        signedPreKey = KeyHelper.generateSignedPreKey(ourIdent, signedKeyId)
        store.storeSignedPreKey(signedPreKey.getId(), signedPreKey)
        result['signedPreKey'] = {
            "keyId": signedPreKey.getId(),
            "publicKey": signedPreKey.getKeyPair().getPublicKey().serialize(),
            "signature": signedPreKey.getSignature()
        }
        store.removeSignedPreKey(signedKeyId - 2)
        store.putState('maxPreKeyId', startId + count)
        store.putState('signedKeyId', signedKeyId + 1)
        return result

    def authHeader(self, username, password):
        return 'Basic ' + b64encode(f'{username}:{password}'.encode()).decode()

    async def request(self, call=None, urn='', method='GET',
                      json=None, username=None, password=None):
        path = SIGNAL_URL_CALLS.get(call) + urn
        headers = {}
        if username and password:
            headers['Authorization'] = self.authHeader(username, password)
        async with self.fetch(path, method=method, json=json,
                              headers=headers) as r:
            is_json = r.content_type.startswith('application/json')
            resp_content = await r.json() if is_json else await r.text()
        # Can we just use native exceptions from aiohttp?? please
        #if not resp.ok:
        #    e = errors.ProtocolError(resp.status, resp_content)
        #    if e.code in SIGNAL_HTTP_MESSAGES:
        #        e.message = SIGNAL_HTTP_MESSAGES[e.code]
        #    else:
        #        e.message = f'Status code: {e.code}'
        #    raise e
        return resp_content

    def fetch(self, urn, headers=None, **kwargs):
        """ Thin wrapper to augment json and auth support. """
        if headers is None:
            headers = {}
        if 'Authorization' not in headers and self.username and self.password:
            headers['Authorization'] = self.authHeader(self.username,
                                                       self.password)
        return self._httpClient.request(url=f'{self.url}/{urn.lstrip("/")}',
                                        headers=headers, **kwargs)

    async def getDevices(self):
        data = await self.request(call='devices')
        return data and data['devices']

    async def registerKeys(self, keys):
        json = {}
        json['identityKey'] = b64encode(keys['identityKey']).decode()
        json['signedPreKey'] = {
            "keyId": keys['signedPreKey']['keyId'],
            "publicKey": b64encode(keys['signedPreKey']['publicKey']).decode(),
            "signature": b64encode(keys['signedPreKey']['signature']).decode()
        }
        json['preKeys'] = [{
            "keyId": pk['keyId'],
            "publicKey": b64encode(pk['publicKey']).decode()
        } for pk in keys['preKeys']]
        return await self.request(call='keys', method='PUT', json=json)

    async def getMyKeys(self):
        res = await self.request(call='keys')
        return res.count

    async def getKeysForAddr(self, addr, device_id='*'):
        res = await self.request(call='keys',
                                 urn=f'/{addr}/{device_id}')
        res['identityKey'] = b64decode(res['identityKey'])
        for device in res['devices']:
            if device['preKey']:
                device['preKey']['publicKey'] = b64decode(device['preKey']['publicKey'])
            device['signedPreKey']['publicKey'] = b64decode(device['signedPreKey']['publicKey'])
            device['signedPreKey']['signature'] = b64decode(device['signedPreKey']['signature'])
        return res

    async def sendMessages(self, destination, messages, timestamp):
        return await self.request(call='messages', method='PUT',
                                  urn='/' + destination,
                                  json={"messages": messages,
                                        "timestamp": timestamp})

    async def getAttachment(self, id):
        """ XXX Build in retry handling... """
        ptr_resp = await self.request(call='attachment', urn='/' + id)
        async with self._httpClient.get(ptr_resp['location']) as r:
            return await r.read()

    async def putAttachment(self, body):
        """ XXX Build in retry handling... """
        ptr_resp = await self.request(call='attachment')
        # Extract the id as a string from the location url
        # (workaround for ids too large for Javascript numbers) # XXX
        import pdb;pdb.set_trace()
        match = self.attachment_id_regex.match(ptr_resp['location'])
        if not match:
            logger.error('Invalid attachment url for outgoing message',
                          ptr_resp['location'])
            raise TypeError('Received invalid attachment url')
        async with self._httpClient.put(ptr_resp['location'], data=body):
            pass
        return match[1]

    def getMessageWebSocketUrl(self):
        return ''.join([
            self.url.replace('https://', 'wss://').replace('http://', 'ws://'),
            f'/v1/websocket/?login={self.username}&password={self.password}'
        ])

    def getProvisioningWebSocketUrl(self):
        return self.url.replace('https://', 'wss://').replace('http://', 'ws://') + \
                                '/v1/websocket/provisioning/'

    async def updateGcmRegistrationId(self, gcm_reg_id):
        """ The GCM reg ID configures the data needed for the PushServer to
        wake us up using google cloud messaging's Push Server (an exercise for
        the user). """
        return await self.request(call='accounts', method='PUT', urn='/gcm',
            json={"gcmRegistrationId": gcm_reg_id})
